### PR TITLE
Use paint for Line vertical scrolling

### PR DESCRIPTION
**Pull request by shtirlic**
Due to unknown reasons, paintLine does not play nice with pseudo/cyr chars, using Paint fixes this.

image

### CODE REVIEW

**Comment by shtirlic** [comment]
@vitaliy-aksyonov  any ideas of the root case?

**Comment by vitaliy-aksyonov** [comment]
What is your environment? Which locale do you use? I have no issues with that when run in ru_RU.KOI8_R. All pseudographics displayed correctly.

**Comment by shtirlic** [comment]
```
+ CMD='/home/shtirlic/bin/golded -VV'
+ export LC_ALL=ru_RU.KOI8-R
+ LC_ALL=ru_RU.KOI8-R
+ export LANG=ru_RU.KOI8-R
+ LANG=ru_RU.KOI8-R
```
Konsole

**Comment by shtirlic** [comment]
Also see this note in advanced config
```
// All pseudo-graphics symbols is replaced with ASCII symbols to
//  prevent garbage displaing of the tables & etc. This table may be used
// on Linux Xterm
;XlatCharSet CP866       KOI8-R      866_koiX.chs

```

**Comment by vitaliy-aksyonov** [comment]
> Also see this note in advanced config
> 
> ```
> // All pseudo-graphics symbols is replaced with ASCII symbols to
> //  prevent garbage displaing of the tables & etc. This table may be used
> // on Linux Xterm
> ;XlatCharSet CP866       KOI8-R      866_koiX.chs
> ```
> 
> maybe you are using this?

Sorry for late reply. I don't use that conversion table and never seen such behavior. No matter if I use ssh and putty or linux console directly.
Could you please provide minimum config and steps to reproduce this issue?

**Comment by vitaliy-aksyonov** [comment]
This have to be fixed by #80
Could you please test if it helps?

**Comment by shtirlic** [comment]
@vitaliy-aksyonov  the same
see the pics

image
and after  i Hit the key down
image
the same with cyr text
image

```
LANG=ru_RU.KOI8-R
LC_CTYPE="ru_RU.KOI8-R"
LC_NUMERIC="ru_RU.KOI8-R"
LC_TIME="ru_RU.KOI8-R"
LC_COLLATE="ru_RU.KOI8-R"
LC_MONETARY="ru_RU.KOI8-R"
LC_MESSAGES="ru_RU.KOI8-R"
LC_PAPER="ru_RU.KOI8-R"
LC_NAME="ru_RU.KOI8-R"
LC_ADDRESS="ru_RU.KOI8-R"
LC_TELEPHONE="ru_RU.KOI8-R"
LC_MEASUREMENT="ru_RU.KOI8-R"
LC_IDENTIFICATION="ru_RU.KOI8-R"
LC_ALL=ru_RU.KOI8-R
```

```
─│┌┐└┘├┤┬┴┼▀▄█▌▐░▒▓⌠■∙√≈≤≥ ⌡°²·÷═║╒ё╓╔╕╖╗╘╙╚╛╜╝╞╟╠╡Ё╢╣╤╥╦╧╨╩╪╫╬©юабцдефгхийклмнопярстужвьызшэщчъЮАБЦДЕФГХИЙКЛМНОПЯРСТУЖВЬЫЗШЭЩЧЪ
```

```
linux-vdso.so.1 (0x00007fbfc2623000)
libncursesw.so.6 => /usr/lib/libncursesw.so.6 (0x00007fbfc255c000)
libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007fbfc1e00000)
libm.so.6 => /usr/lib/libm.so.6 (0x00007fbfc2471000)
libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007fbfc21d3000)
libc.so.6 => /usr/lib/libc.so.6 (0x00007fbfc1c14000)
/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fbfc2625000)
```

**Comment by shtirlic** [comment]
Is there any big performance benefits to paint by line and do some crazy stuff or instead just use repaint all for simplify ?

**Comment by vitaliy-aksyonov** [comment]
I personally don't like fixes which doesn't fix root cause. Do you know why your change help and why original code works incorrectly?

Could you please run this code snippet on your computer and send me results?

`#include <locale>
#include <iostream>

void checks(unsigned c)
{
    std::cout << "Char: " << c <<  " isprint: " << isprint((int)c)
    << "\n iscntrl: " << iscntrl((int) c) << std::endl;
}

int main()
{
    setlocale(LC_ALL, "");
    unsigned c = 0x80;
    do
    {
        checks(c);
    } while (c++ < 0xff);
}`

All those symbols shall have isprint true and iscntrl false.

File: golded3/geview.cpp [hunk 1]
```diff
@@ -498,53 +498,49 @@ void GMsgBodyView::Use(Area *areaptr, GMsg *msgptr, int startline)
 
 //  ------------------------------------------------------------------
 
-void GMsgBodyView::PaintLine(int row, Line *line)
-{
-
+void GMsgBodyView::PaintLine(int row, Line *line) {
```
Comment (vitaliy-aksyonov) [inline golded3/geview.cpp:501]: Don't change code style please. Open curly brace shall be on new line.

File: golded3/geview.cpp [hunk 2]
```diff
@@ -498,53 +498,49 @@ void GMsgBodyView::Use(Area *areaptr, GMsg *msgptr, int startline)
...
 
-    // Trim line if it longer than should be. This actually happens in very rare
-    // cases, but always when hex dump displayed.
-    if(llen > visible_width)
-    {
-        llen = visible_width;
-        line->txt.erase(llen);
+    // Trim line if it is longer than it should be
+    if (textLength > visible_width) {
+        text = text.substr(0, visible_width);
```
Comment (vitaliy-aksyonov) [inline golded3/geview.cpp:511]: substr() creates new string. erase() is more efficient. Leave it as it was.

File: golded3/geview.cpp [hunk 3]
```diff
@@ -498,53 +498,49 @@ void GMsgBodyView::Use(Area *areaptr, GMsg *msgptr, int startline)
...
 
-void GMsgBodyView::Paint()
-{
+//  ------------------------------------------------------------------
 
+void GMsgBodyView::Paint() {
     window.activate_quick();
 
-    Line* dummy_index = NULL;
+    Line* dummy_index = nullptr;
```
Comment (vitaliy-aksyonov) [inline golded3/geview.cpp:537]: This won't compile in older compilers. `nullptr` is C++11.

**Review by vitaliy-aksyonov — CHANGES_REQUESTED** [review]

**Comment by shtirlic** [comment]
@vitaliy-aksyonov thx, actually I think I found the issue and now researching looks like code related to window scroll and in the way how `vsave` and `vrestore` works via curses, I will update when I have some more info.

**Comment by vitaliy-aksyonov** [comment]
It could be. My build uses non wide ncurses.

**Comment by shtirlic** [comment]
> It could be. My build uses non wide ncurses.

Pls, could you try build with ncurses wide support  WIDE_NCURSES=1, maybe you will also encounter this bug.

**Comment by vitaliy-aksyonov** [comment]
Sure. I'll try later.

**Comment by shtirlic** [comment]
Closing this because root case was found and should be addressed via  #89

**Pull request closed without merging**